packages/js/src/engine.ts

```diff
@@ -175,3 +175,19 @@ export function matchProvider(providers: Provider[], { modelId, providerApiUrl,
 export function matchModel(models: ModelInfo[], modelId: string): ModelInfo | undefined {
   return models.find((m) => matchLogic(m.match, modelId))
 }
+
+export function matchModelWithFallback(provider: Provider, modelId: string, allProviders: Provider[]): ModelInfo | undefined {
+  const model = matchModel(provider.models, modelId)
+  if (model) return model
+
+  if (provider.fallback_model_providers) {
+    for (const fallbackProviderId of provider.fallback_model_providers) {
+      const fallbackProvider = allProviders.find((p) => p.id === fallbackProviderId)
+      if (fallbackProvider) {
+        const fallbackModel = matchModelWithFallback(fallbackProvider, modelId, allProviders)
```

Comment:
Is this recursion intentional? I'm a bit uncomfortable with it, not sure if providers can't cross-reference each other.